### PR TITLE
ascanrulesAlpha: Add SSRF Scan Rule

### PR DESCRIPTION
**Pull request by ricekot**
Closes zaproxy/zaproxy#7360.

### CODE REVIEW

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                Control.getSingleton().getExtensionLoader().getExtension(ExtensionOast.class);
+        try {
+            HttpMessage msg = getNewMsg();
+            Alert alert =
+                    newAlert()
+                            .setConfidence(Alert.CONFIDENCE_MEDIUM)
+                            .setMessage(msg)
+                            .setSource(Alert.Source.ACTIVE)
+                            .build();
+            String[] schemes = {"http", "https"};
```
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java:158]: I believe there are constants for the in the core HttpHeader classes

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            for (String scheme : schemes) {
+                OastPayload oastPayload = extOast.registerAlertAndGetOastPayload(alert);
+                String payload = scheme + "://" + oastPayload.getPayload();
+                alert.setParam(param);
```
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java:162]: I believe there are also Constants that include colon slash slash

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.core.scanner.AbstractAppParamPlugin;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Category;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.addon.commonlib.CommonAlertTag;
+import org.zaproxy.addon.oast.ExtensionOast;
+import org.zaproxy.addon.oast.OastPayload;
+
+public class SsrfScanRule extends AbstractAppParamPlugin {
```
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java:37]: Worth adding the example alerts method.

File: addOns/ascanrulesAlpha/ascanrulesAlpha.gradle.kts [hunk 4]
```diff
@@ -14,7 +14,7 @@ zapAddOn {
                     version.set(">= 1.9.0 & < 2.0.0")
                 }
                 register("oast") {
-                    version.set(">= 0.7.0")
+                    version.set(">= 0.13.0")
```
Comment (thc202) [inline addOns/ascanrulesAlpha/ascanrulesAlpha.gradle.kts:17]: Should be 0.14.0.

**Review by kingthorin — APPROVED** [review]

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                                CommonAlertTag.OWASP_2021_A10_SSRF,
+                                CommonAlertTag.WSTG_V42_INPV_19_SSRF));
+        alertTags.put(ExtensionOast.OAST_ALERT_TAG_KEY, ExtensionOast.OAST_ALERT_TAG_VALUE);
+        return alertTags;
+    }
+
+    @Override
+    public List<Alert> getExampleAlerts() {
+        try {
+            HttpMessage msg = new HttpMessage(new URI("https://example.com", true));
```
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java:100]: This is not needed, building the Alert does not require a message.

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            return List.of(
+                    newAlert()
+                            .setConfidence(Alert.CONFIDENCE_MEDIUM)
+                            .setMessage(msg)
+                            .setSource(Alert.Source.ACTIVE)
```
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/SsrfScanRule.java:105]: Other scan rules are not setting it.

**Comment by thc202** [comment]
Would be good to have tests for both the scan rule and the OAST changes (not a must though).

**Comment by roisec** [comment]
@ricekot ,
I am trying to understand why my issue regarding the SSRF rule closed(https://github.com/zaproxy/zaproxy/issues/7360), and now you decided to continue and build the SSRF rule.

**Comment by thc202** [comment]
The issue is still open.

**Comment by roisec** [comment]
Is this rule injected for every get/post/JSON parameters value and link to the OAST server?

**Comment by thc202** [comment]
Yes.

**Comment by roisec** [comment]
@thc202 Thanks!
Waiting to check this rule.

**Review by thc202 — APPROVED** [review]

**Review by kingthorin — APPROVED** [review]

**Pull request merged**